[quantization] Dynamic dispatch for optimized quantized op kernels

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25545 [quantization] Dynamic dispatch for optimized quantized op kernels**

This re-uses the infrastructure from ATen/native/cpu, which compiles kernels multiple times for different instruction sets and dispatches dynamically based on the CPU's capability flags at runtime. This ensures we use the most optimal quantized kernel for the given machine

Differential Revision: [D17166369](https://our.internmc.facebook.com/intern/diff/D17166369)

Commit:
[quantization] Dynamic dispatch for optimized quantized op kernels

Commit:
Update on "[quantization] Dynamic dispatch for optimized quantized op kernels"


This re-uses the infrastructure from ATen/native/cpu, which compiles kernels multiple times for different instruction sets and dispatches dynamically based on the CPU's capability flags at runtime. This ensures we use the most optimal quantized kernel for the given machine

Commit:
Update on "[quantization] Dynamic dispatch for optimized quantized op kernels"


This re-uses the infrastructure from ATen/native/cpu, which compiles kernels multiple times for different instruction sets and dispatches dynamically based on the CPU's capability flags at runtime. This ensures we use the most optimal quantized kernel for the given machine

VitalyFedyunin commented on cmake/Codegen.cmake:
```diff
@@ -140,6 +140,25 @@ if (INTERN_BUILD_ATEN_OPS)
 
+  FILE(GLOB cpu_quantized_kernel_cpp_in "${CMAKE_CURRENT_LIST_DIR}/../aten/src/ATen/native/quantized/cpu/kernels/*.cpp")
+  FOREACH(i RANGE ${NUM_CPU_CAPABILITY_NAMES})
+    FOREACH(IMPL ${cpu_quantized_kernel_cpp_in})
```
Instead of copy-paste can you combine `cpu_kernel_cpp_in` and `cpu_kernel_cpp_in` to iterate them together.

It is prote to error otherwise as you already (likely) adding files twice, but not cleaning `cpu_kernel_cpp` content between loops.

VitalyFedyunin commented on aten/src/ATen/native/quantized/cpu/qrelu.cpp:
```diff
@@ -51,32 +39,10 @@ namespace {
 Tensor quantized_relu6(const Tensor& qx) {
   Tensor qy;
   const auto zero_point = qx.q_zero_point();
```
unused

VitalyFedyunin reviewed (commented):
Overal LGTM with couple small remarks.

Commit:
Update on "[quantization] Dynamic dispatch for optimized quantized op kernels"


This re-uses the infrastructure from ATen/native/cpu, which compiles kernels multiple times for different instruction sets and dispatches dynamically based on the CPU's capability flags at runtime. This ensures we use the most optimal quantized kernel for the given machine

Differential Revision: [D17166369](https://our.internmc.facebook.com/intern/diff/D17166369)

raghuramank100 commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -405,8 +405,8 @@ struct Vec256<c10::qint32> {
-      return _mm256_min_epi8(
-          _mm256_max_epi8(vals, zero_point.vals), q_six.vals);
+      return _mm256_min_epi32(
+          _mm256_max_epi32(vals, zero_point.vals), q_six.vals);
```
For int32, the only value of zero-point supported is 0, can we add an explicit check for that?

jamesr66a replied:
Not here, no, this is used in the tight loop of the inner kernel

Why do you want this check? Correctness? It wouldn't help performance at all

raghuramank100 replied:
Not in the kernel, do we have a check for relu/relu6 that checks for this? The place to do this would be right the beginning of the function call: https://github.com/pytorch/pytorch/pull/25545/files#diff-a04b434d724c12b848bd2bc5d8dadd40R15

jamesr66a replied:
That check doesn't belong there either, as mentioned in the README that file is only for things that differ depending on the CPU architecture. We can add this check to e.g. `qrelu.cpp` but I don't think this is relevant for this PR

Commit:
Update on "[quantization] Dynamic dispatch for optimized quantized op kernels"


This re-uses the infrastructure from ATen/native/cpu, which compiles kernels multiple times for different instruction sets and dispatches dynamically based on the CPU's capability flags at runtime. This ensures we use the most optimal quantized kernel for the given machine

Differential Revision: [D17166369](https://our.internmc.facebook.com/intern/diff/D17166369)

Commit:
Update on "[quantization] Dynamic dispatch for optimized quantized op kernels"


This re-uses the infrastructure from ATen/native/cpu, which compiles kernels multiple times for different instruction sets and dispatches dynamically based on the CPU's capability flags at runtime. This ensures we use the most optimal quantized kernel for the given machine

Differential Revision: [D17166369](https://our.internmc.facebook.com/intern/diff/D17166369)

raghuramank100 reviewed (approved):
LGTM
